### ISSUE TITLE
[Bug]: --enable-lora causes premature response truncation and wrong stop token (<|endoftext|> instead of <|im_end|>) with Qwen3

### ISSUE THREAD

**Issue by 18174045485** [vllm-project/vllm-ascend#8863]
## Environment Information

Environment was collected with the vLLM Ascend environment collection script.

### Container Image

```text
m.daocloud.io/quay.io/ascend/vllm-ascend:v0.14.0rc1-openeuler
```

### vLLM / vLLM Ascend Versions

```text
vLLM Version: 0.14.1
vLLM Ascend Version: 0.14.0rc1
```

`pip list` also shows:

```text
vllm          0.14.1+empty
vllm-ascend   0.14.0rc1
```

### PyTorch / Python

```text
PyTorch version: 2.9.0+cpu
Is debug build: False

Python version: 3.11.14
Python platform: Linux-4.19.90-23.62.v2101.ky10.aarch64-aarch64-with-glibc2.38
```

### Operating System

```text
OS: openEuler 24.03 (LTS-SP2) (aarch64)
Libc version: glibc-2.38
```

### Compiler / Build Tools

```text
GCC version: 12.3.1
Clang version: 17.0.6
CMake version: 4.2.1
```

### Relevant Python Packages

```text
numpy==1.26.4
pyzmq==27.1.0
sentence-transformers==5.4.1
torch==2.9.0+cpu
torch_npu==2.9.0
torchvision==0.24.0
transformers==4.57.6
triton-ascend==3.2.0
```

### CPU

```text
Architecture: aarch64
Vendor ID: HiSilicon
Model name: Kunpeng-920
CPU(s): 192
Thread(s) per core: 1
Core(s) per cluster: 48
Cluster(s): 4
NUMA node(s): 8
CPU max MHz: 2600.0000
CPU min MHz: 200.0000
```

### NPU

```text
npu-smi version: 25.5.1
NPU count: 8
NPU model: 910B3
Health: OK for all devices
HBM per device: 65536 MB
```

Detected devices:

```text
NPU 0: 910B3, Bus-Id 0000:C1:00.0, HBM 3407 / 65536 MB
NPU 1: 910B3, Bus-Id 0000:C2:00.0, HBM 3398 / 65536 MB
NPU 2: 910B3, Bus-Id 0000:81:00.0, HBM 3400 / 65536 MB
NPU 3: 910B3, Bus-Id 0000:82:00.0, HBM 3398 / 65536 MB
NPU 4: 910B3, Bus-Id 0000:01:00.0, HBM 3400 / 65536 MB
NPU 5: 910B3, Bus-Id 0000:02:00.0, HBM 3396 / 65536 MB
NPU 6: 910B3, Bus-Id 0000:41:00.0, HBM 3398 / 65536 MB
NPU 7: 910B3, Bus-Id 0000:42:00.0, HBM 3397 / 65536 MB
```

No running NPU processes were found when the environment information was collected.

### CANN

```text
package_name=Ascend-cann-toolkit
version=8.5.0
innerversion=V100R001C25SPC001B232
arch=aarch64
os=linux
path=/usr/local/Ascend/cann-8.5.0
```

### Ascend-related Environment Variables

```text
ASCEND_TOOLKIT_LATEST_HOME=/usr/local/Ascend/ascend-toolkit/latest
ASCEND_TOOLKIT_HOME=/usr/local/Ascend/cann-8.5.0
ASCEND_HOME_PATH=/usr/local/Ascend/cann-8.5.0
ASCEND_OPP_PATH=/usr/local/Ascend/cann-8.5.0/opp
ASCEND_AICPU_PATH=/usr/local/Ascend/cann-8.5.0
ATB_HOME_PATH=/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_1
TORCH_DEVICE_BACKEND_AUTOLOAD=1
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
```

### ATB Environment Variables

```text
ATB_OPSRUNNER_KERNEL_CACHE_LOCAL_COUNT=1
ATB_OPSRUNNER_KERNEL_CACHE_GLOABL_COUNT=5
ATB_STREAM_SYNC_EVERY_RUNNER_ENABLE=0
ATB_STREAM_SYNC_EVERY_KERNEL_ENABLE=0
ATB_STREAM_SYNC_EVERY_OPERATION_ENABLE=0
ATB_COMPARE_TILING_EVERY_KERNEL=0
ATB_MATMUL_SHUFFLE_K_ENABLE=1
ATB_WORKSPACE_MEM_ALLOC_ALG_TYPE=1
ATB_SHARE_MEMORY_NAME_SUFFIX=
```

### LD_LIBRARY_PATH

```text
LD_LIBRARY_PATH includes:

/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_1/lib
/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_1/examples
/usr/local/Ascend/nnal/atb/latest/atb/cxx_abi_1/tests/atbopstest
/usr/local/Ascend/cann-8.5.0/lib64
/usr/local/Ascend/cann-8.5.0/lib64/plugin/opskernel
/usr/local/Ascend/cann-8.5.0/lib64/plugin/nnengine
/usr/local/Ascend/cann-8.5.0/opp/built-in/op_impl/ai_core/tbe/op_tiling/lib/linux/aarch64
/usr/local/Ascend/cann-8.5.0/tools/aml/lib64
/usr/local/Ascend/cann-8.5.0/tools/aml/lib64/plugin
/usr/local/Ascend/driver/lib64
/usr/local/Ascend/driver/lib64/common
/usr/local/Ascend/driver/lib64/driver
/usr/local/Ascend/ascend-toolkit/latest/tools/aml/lib64
/usr/local/Ascend/ascend-toolkit/latest/tools/aml/lib64/plugin
/usr/local/Ascend/ascend-toolkit/latest/lib64
/usr/local/Ascend/ascend-toolkit/latest/lib64/plugin/opskernel
/usr/local/Ascend/ascend-toolkit/latest/lib64/plugin/nnengine
/usr/local/Ascend/ascend-toolkit/latest/opp/built-in/op_impl/ai_core/tbe/op_tiling
```

## Bug Description

When vLLM is started with `--enable-lora`, even without any `--lora-modules`, both streaming and non-streaming chat completions exhibit multiple abnormal behaviors:

- Premature truncation
- Wrong stop token
- Word repetitions
- Reasoning parser failure

Removing `--enable-lora` restores correct behavior in all cases.

## Minimal Reproduction

Start vLLM with `--enable-lora --enforce-eager`, but without any `--lora-modules`:

```bash
python -m vllm.entrypoints.openai.api_server \
    --model /workspace/Qwen3-32B \
    --tensor-parallel-size 2 \
    --reasoning-parser qwen3 \
    --enable-lora \
    --enforce-eager
```

Send a non-streaming request:

```bash
curl http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "/workspace/Qwen3-32B",
    "messages": [{"role": "user", "content": "圆周率是什么"}],
    "stream": false,
    "max_tokens": 2048
  }'
```

## Expected Behavior

The response should complete normally:

- `finish_reason`: `"stop"`
- `stop_reason`: `151645`, corresponding to `<|im_end|>`, the correct chat end token for Qwen3
- `reasoning_content` should be correctly populated
- Raw `<think>` tags should not appear in `message.content`
- Streaming and non-streaming responses should both complete without premature truncation

## Actual Behavior

Three different failure modes are observed across requests.

### Mode 1: Wrong stop token, truncated output, and word repetitions

Example response:

```json
{
  "finish_reason": "stop",
  "stop_reason": 151643,
  "message": {
    "content": "...尽管\n",
    "reasoning_content": "...性质的性质和性质，然后..."
  }
}
```

Observed symptoms:

- Response is truncated mid-sentence.
- `stop_reason` is `151643`, corresponding to `<|endoftext|>`.
- For Qwen3 chat completions, the expected chat stop token is `151645`, corresponding to `<|im_end|>`.
- Word repetitions appear in both `message.content` and `reasoning_content`.

Examples of repetitions:

```text
重要重要
出现在出现在
性质的性质
```

### Mode 2: Reasoning parser failure and raw `<think>` tags in `content`

Example response:

```json
{
  "finish_reason": "stop",
  "stop_reason": null,
  "message": {
    "content": "<think>\n嗯，用户问\"圆周率是什么\"...\n\n圆周率（",
    "reasoning_content": null
  },
  "usage": {
    "prompt_tokens": 12,
    "completion_tokens": 63
  }
}
```

Observed symptoms:

- `reasoning_content` is `null`.
- Thinking content appears as raw `<think>` tags inside `message.content`.
- `--reasoning-parser qwen3` appears to be silently bypassed or broken.
- Response is again truncated mid-sentence.

### Mode 3: Streaming truncation

Streaming responses with `"stream": true` are also truncated prematurely, apparently with the same root cause.

## Summary of Observed Symptoms

Without `--enable-lora`, all behaviors are normal:

- Responses complete fully.
- `stop_reason` is `151645`, corresponding to `<|im_end|>`.
- `reasoning_content` is correctly populated.
- Raw `<think>` tags do not appear in `message.content`.
- Streaming works without interruption.

With `--enable-lora`, the generation pipeline appears to break down:

- Responses are consistently truncated mid-sentence.
- `stop_reason` is either `151643`, corresponding to `<|endoftext|>`, or `null`.
- Neither `151643` nor `null` is the correct chat stop token for Qwen3.
- `--reasoning-parser qwen3` stops functioning.
- `reasoning_content` becomes `null`.
- Raw `<think>` tags appear inside `message.content`.
- Word repetitions appear in the generated text, for example `重要重要`, `出现在出现在`, and `性质的性质`.
- These failures occur in both streaming and non-streaming modes.

## Key Observations

- Without `--enable-lora`, all of the above behaviors work correctly.
- With `--enable-lora` and no `--lora-modules`, all symptoms appear.
- This suggests the LoRA infrastructure itself is the trigger, not the LoRA weights.
- With `--enable-lora --lora-modules probe=<path>`, the same failures are observed.
- The issue is reproducible on every request and does not appear to be intermittent.

## Hypothesis

On vLLM Ascend NPU, enabling `--enable-lora` appears to replace or alter a core component in the token sampling, logits processing, or stop-token handling pipeline, causing:

1. Incorrect stop token selection, using `<|endoftext|>` instead of `<|im_end|>`
2. Reasoning parser bypass or failure
3. Abnormal repetition in generated tokens
4. Premature truncation in both streaming and non-streaming responses

The key point is that the issue reproduces even when `--enable-lora` is set without any `--lora-modules`, which suggests that the problem is triggered by enabling the LoRA code path itself.

## Additional Note

The above analysis was assisted by Claude Sonnet 4.6.

**Comment by linfeng-yuan** [comment]
Hi, thanks for your detailed feedback. I notice that there are some bug fix commits in upstream vllm and vllm-ascend aroud `max_loras` and `num_active_lora` (e.g., https://github.com/vllm-project/vllm-ascend/pull/6958). It is recommended to use vllm and vllm-ascend with `v0.18.0` tags. You can also check this e2e test case which is nearly same as your case: https://github.com/vllm-project/vllm-ascend/blob/main/tests/e2e/multicard/2-cards/test_llama32_lora_tp2.py

Additionally, we would assign this problem to @ZYang6263 . Please have a look.
